File: backend/app/schemas/template.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel


class TemplateBase(BaseModel):
    name: str
    description: Optional[str] = None
    icon: Optional[str] = None
    category: Optional[str] = None
    template_title: Optional[str] = None
    template_description: Optional[str] = None
    template_priority: Optional[int] = None
    template_project_id: Optional[int] = None
    template_tags_json: Optional[str] = None
    template_recurrence_json: Optional[str] = None
    template_duration: Optional[int] = None
    template_subtasks_json: Optional[str] = None


class TemplateCreate(TemplateBase):
    pass


class TemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    icon: Optional[str] = None
    category: Optional[str] = None
    template_title: Optional[str] = None
    template_description: Optional[str] = None
    template_priority: Optional[int] = None
    template_project_id: Optional[int] = None
    template_tags_json: Optional[str] = None
    template_recurrence_json: Optional[str] = None
    template_duration: Optional[int] = None
    template_subtasks_json: Optional[str] = None


class TemplateResponse(TemplateBase):
    id: int
    user_id: int
    is_built_in: bool
    usage_count: int
    last_used_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TemplateUseRequest(BaseModel):
    due_date: Optional[date] = None
    project_id: Optional[int] = None


class TemplateUseResponse(BaseModel):
    task_id: int
    message: str


class TemplateFromTaskRequest(BaseModel):
    name: str
    icon: Optional[str] = None
    category: Optional[str] = None
    description: Optional[str] = None

File: web/src/types/template.ts
export interface TaskTemplate {
  id: string;
  user_id: string;
  name: string;
  description: string | null;
  icon: string | null;
  category: string | null;
  template_title: string | null;
  template_description: string | null;
  template_priority: number | null;
  template_project_id: string | null;
  template_tags_json: string | null;
  template_recurrence_json: string | null;
  template_duration: number | null;
  template_subtasks_json: string | null;
  is_built_in: boolean;
  usage_count: number;
  last_used_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface TemplateCreate {
  name: string;
  description?: string;
  icon?: string;
  category?: string;
  template_title?: string;
  template_description?: string;
  template_priority?: number;
  template_project_id?: string;
  template_tags_json?: string;
  template_recurrence_json?: string;
  template_duration?: number;
  template_subtasks_json?: string;
}

export interface TemplateUpdate {
  name?: string;
  description?: string;
  icon?: string;
  category?: string;
  template_title?: string;
  template_description?: string;
  template_priority?: number;
  template_project_id?: string;
  template_tags_json?: string;
  template_recurrence_json?: string;
  template_duration?: number;
  template_subtasks_json?: string;
}

export interface TemplateUseRequest {
  due_date?: string;
  project_id?: string;
}

export interface TemplateUseResponse {
  task_id: string;
  message: string;
}
